### PR TITLE
Strengthen broken infinite scroll in some browsers and site configurations

### PR DESCRIPTION
**Pull request by houmark**
**Problem 1:**
Some browsers (among them Safari) fails to find the scroll position from html / body using jQuery. This is potentially related to the site having `overflow-x: hidden;` on the `html`tag.

This PR strengthens the browser lookup for the scroll position in all browsers (that I was able to test).

**Problem 2:**
Additionally jQuery fails to find the `offset.top` of the hidden "load more" button, due the button being `display: none;` in CSS/LESS. According to the jQuery documentation an element being styled with `display: none;` is excluded from the rendering tree and thus has a position that is undefined. This has been changed to `display:block;` and `visibility:hidden` This can affect the invisible space on existing sites below the AJAX content, but there's no other way and site devs can work around this by decreasing the height of the element and/or its parent. The button and it's parent(s) is removed after all content has been added to the page (nothing changed here).

**Additional improvements:**

* The on scroll event was always attached, even on pages which did not have any AJAX features. This PR does a check for the `data-apos-ajax-infinite-scroll` and only attaches the event watcher in case there is such an element in the DOM.

* The aforementioned scroll event watcher was firing on every scroll and this can lead to significant slugginess on scroll, especially on devices with less power, such as mobile. I have wrapped the event in a lodash throttle function which ensures a scroll event is only sent every 200ms. This should be plenty, as the code operates with a decent offset (500px default). In my testing I did not notice any delays in fetching data.

There's more improvements to be done here, but small steps (new around here). At least this is now fully working for me (it was broken out of the box in all browsers due to both main problems).

### CODE REVIEW

**Comment by crstnio** [comment]
Love it! How about setting the `height` of this button to `0`? Would that remove the whitespace and work with jQuery at the same time?

**Comment by houmark** [comment]
@crstnio I think it will.

When I wrote the long commit message, I was considering to do a test on that, but decided to commit it in and see what you guys think.

At the same time maybe Apostrophe should not be too opinionated about this part, and documentation could simply mention that one can overwrite styling and size down the button, if it's important for the site design (the docs for this parts needs a bit of clarification in the parts anyways, something I may push PR for in the near future).

**Comment by houmark** [comment]
Btw, `$._data(window, "events");` used in Console to check if the scroll listener was correctly attached (or not attached when not wanted). Just in case someone needs it :)

**Comment by crstnio** [comment]
Hmm, I think this fix should not _break_ existing layout by adding unnecessary white space below AJAX content. The developer may not find immediately what is causing the additional space. That's not clean, at least for me. :)

**Comment by houmark** [comment]
The main thing to keep in mind here is that it’s broken for all users right now. At least in my testing Chrome was feeding all records consistently because it couldn’t find the button offset. You may not notice it, or it may work different it in another design than the one I tested in.

What’s complicating things here are that Apostophe is removing any parent wrapper around the button after the last items have loaded (that may or may not be intended, not sure). And so when you get to the sibling element below, and it stops scrolling and adding items, the button element is fully gone from the DOM, also with my changes. So it’s only potentially adding more space below the area of the items while it’s still loading items, so you probably wouldn’t ever notice it.

**Comment by boutell** [comment]
@stuartromanek Thoughts? This looks good to me.

File: lib/modules/apostrophe-ui/public/js/ui.js [hunk 1]
```diff
@@ -432,13 +432,15 @@ apos.define('apostrophe-ui', {
       });
 
       $(function() {
-        self.ajaxInfiniteScrollHandler();
-      });
-
-      $(window).on('scroll', function(event) {
-        self.ajaxInfiniteScrollHandler();
+        var $infiniteScroll = $('[data-apos-ajax-infinite-scroll]');
```
[the remark this answers is not in the document]
Comment (boutell) [reply lib/modules/apostrophe-ui/public/js/ui.js]: Actually, infinite scroll elements could be added by AJAX, and the current code accommodates that.

**Comment by houmark** [comment]
I'm a bit confused now, so I would like your feedback if any changes are needed at this time or not.

Infinite scroll can be added by AJAX, I did not test this, and that's a valid point. I would need to do a test case with that scenario to proper test, but I would prefer to try to attach the `ajaxInfiniteScrollHandler` when content is added by AJAX.

The button is removed when no more pieces exists (else where in the code, not changed in this PR), and that will be handled on line 446-449.

By basics, I don't think Apostrophe should add scroll handlers on ALL pages for something that may only be on some pages or on some sites never. Having a scroll handler around, fast or not, will cost extra CPU cycles on every scroll on every page. It's good practice to throttle very often repeating events like that, and since there's a default offset of 500px, it's very unlikely that anyone scrolls 500px in less than 200ms and if they do, then it will just catch up with a small delay of maximum 200ms

**Comment by houmark** [comment]
After having giving this another thought, I decided to remove the check for infinite scroll elements on the page on load. Considering @boutell's comments, a infinite scroll could be added by JavaScript after the initial page load, this would require a DOM watcher that would enable the on scroll event watcher.

This is replacing one watcher with another one (and the latter is probably much heavier due to Apostrophe being 100% frontend editing).

Since the on scroll watcher is now throttled, it should be very little overhead this creates on scroll of pages.

Considering this one done. Looking forward to see this merged.

**Pull request merged**